src/server/cluster/cluster_family.cc

```diff
@@ -689,6 +696,41 @@ void ClusterFamily::DflySlotMigrationStatus(CmdArgList args, ConnectionContext*
   return rb->SendSimpleString(state_to_str(MigrationState::C_NO_STATE));
 }
 
+void ClusterFamily::DflyClusterMigrationFinalize(CmdArgList args, ConnectionContext* cntx) {
+  CmdArgParser parser{args};
+  auto sync_id = parser.Next<uint32_t>();
+
+  if (auto err = parser.Error(); err) {
+    return cntx->SendError(err->MakeReply());
+  }
+
+  auto migration_it = outgoing_migration_jobs_.find(sync_id);
+  if (migration_it == outgoing_migration_jobs_.end()) {
+    return cntx->SendError(absl::StrCat("finalize invalid session id ", sync_id));
+  }
+
+  if (migration_it->second->GetState() != MigrationState::C_STABLE_SYNC) {
+    return cntx->SendError("Migration process is not in STABLE_SYNC state");
+  }
+
+  shard_set->pool()->AwaitFiberOnAll([migration_it](auto*) {
+    if (const auto* shard = EngineShard::tlocal(); shard)
+      migration_it->second->Finalize(shard->shard_id());
+  });
+
+  // TODO do next after ACK
+  util::ThisFiber::SleepFor(500ms);
+
+  shard_set->pool()->AwaitFiberOnAll([migration_it](auto*) {
+    if (const auto* shard = EngineShard::tlocal(); shard)
+      migration_it->second->Cancel(shard->shard_id());
+  });
```

Comment:
```suggestion
  shard_set->pool()->AwaitFiberOnAll([migration_it](auto*) {
    if (const auto* shard = EngineShard::tlocal(); shard)
      migration_it->second->Finalize(shard->shard_id());

      // TODO do next after ACK
      util::ThisFiber::SleepFor(500ms);

      migration_it->second->Cancel(shard->shard_id());
  });
```

I imagine we will also wait for the ACK in each of the shards, right?

Reply:
no we will have only one ack and I think they will be here or in OutgoingMigration::Finalize

Reply:
>  // TODO do next after ACK
>  util::ThisFiber::SleepFor(500ms);

also TODO: actually stop traffic to make transition atomic 🙂

Reply:
I already in progress with that) so I will not forget about it